### PR TITLE
feat(NewTab): Get items from Pocket Graph [DIS-452]

### PR DESCRIPTION
**Pull request by mmiermans**
# Goal
Get CorpusItems directly from the Pocket Graph, instead of from a Feature Group. This improves reliability for New Tab:

1. There are fewer dependencies in the critical path. The Pocket Graph and Curated Corpus API are already in the critical path to provide the CorpusItem title, excerpt, etc.
2. It avoids a long delay between generating the candidates and decorating them. A recent outage was caused by the data pipeline getting out-of-sync with the Curated Corpus API, which is less likely to happen with this setup.

## Reference

Tickets:
* https://getpocket.atlassian.net/browse/DIS-452

New Tab slate spec:
* https://getpocket.atlassian.net/wiki/spaces/PE/pages/2927100008/Fx+NewTab+Slate+spec+for+New+Markets

## Implementation Decisions
- SlateProvider now accepts an abstract `CorpusFetchable` that fetches corpus items. One implementation fetches from the Feature Group and the one used for New Tab fetches from the Pocket Graph.
- Change the functional test to pytest, and hit the public production Pocket Graph. Reasoning is explained in the test.

### CODE REVIEW

File: app/data_providers/slate_providers/new_tab_slate_provider.py [hunk 1]
```diff
@@ -52,8 +71,7 @@ def _get_tile_id_key(self, item: CorpusItemModel) -> str:
         """
         :return: A string that identifiers the scheduled surface, scheduled date, and CorpusItem.
         """
-        # TODO: When scheduledSurface.items is queried from the Graph, fill in the scheduledDate below. [DIS-452]
-        return f'{self.recommendation_surface_id}/{item.id}/<TODO: pull in ScheduledSurfaceItem.scheduledDate>'
+        return f'{self.recommendation_surface_id}/{item.id}/{self.corpus_api_client.get_scheduled_date(item.id)}'
```
[the remark this answers is not in the document]
Comment (mmiermans) [reply app/data_providers/slate_providers/new_tab_slate_provider.py:74]: It won't fail in that case, because f-strings can take anything that has a string representation, which is pretty much anything in Python.
```python
>>> f'surface/123/{None}'
'surface/123/None'
```

File: app/data_providers/slate_providers/new_tab_slate_provider.py [hunk 2]
```diff
@@ -70,4 +88,7 @@ async def rank_corpus_items(self, items: List[CorpusItemModel], *args, **kwargs)
             default_alpha_prior=188,  # beta * P99 German NewTab CTR for 2023-03-28 to 2023-04-05 (1.5%)
             default_beta_prior=12500)  # 0.5% of median German NewTab item impressions for 2023-03-28 to 2023-04-05.
 
+        # Sort newest to oldest. Python is stable, so it will preserve the Thompson sampling within a scheduled date.
+        items.sort(key=lambda item: self.corpus_api_client.get_scheduled_date(item.id), reverse=True)
```
[the remark this answers is not in the document]
Comment (mmiermans) [reply app/data_providers/slate_providers/new_tab_slate_provider.py:92]: New Tab would in that case get the highest-CTR items from the last two days.

I'll add a unit test to cover the case where scheduledDate is None.

File: tests/functional/graphql/test_new_tab_slate.py [hunk 3]
```diff
@@ -35,49 +25,57 @@ def _format_new_tab_query(locale, region, count=50):
...
-            data = response.json()
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    "locale,region",
+    [
+        ('es-ES', 'ES'),
+        ('fr-FR', 'FR'),
+        ('it-IT', 'IT'),
+    ])
+async def test_new_tab_slate(locale, region, snowplow_micro, pocket_graph_request_headers):
```
[the remark this answers is not in the document]
Comment (mmiermans) [reply tests/functional/graphql/test_new_tab_slate.py:54]: :+1: I'll add a test case where the region parameter (which is optional) is missing.

File: app/data_providers/slate_providers/new_tab_slate_provider.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (eu9ene) [reply app/data_providers/slate_providers/new_tab_slate_provider.py:74]: True, but are we ok with having None there?

File: app/data_providers/slate_providers/new_tab_slate_provider.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmiermans) [reply app/data_providers/slate_providers/new_tab_slate_provider.py:92]: Good thing you mentioned this, because `None < None` does not compute.

File: app/data_providers/slate_providers/new_tab_slate_provider.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mmiermans) [reply app/data_providers/slate_providers/new_tab_slate_provider.py:75]: Good question. Yes, I think that's an acceptable way to degrade in case of a failure.

Here's a possible scenario:
- A bug is introduced in the Curated Corpus API that returns `null` for `scheduledDate`. (This is never expected to happen normally, because we're fetching scheduled items.)
- Recommendation API starts logs an error to Sentry.
- Recommendation API continues to return recommendations. The `tileId` will be different because the date part is missing. - There probably will be no impact at all, but if there is any, it would be some low-impact reporting or Thompson sampling effect. If Thompson sampling would start getting stale data, an alert would go off for the [recommendation freshness SLO](https://getpocket.atlassian.net/wiki/spaces/PE/pages/2927099995/Fx+NewTab+for+New+Markets+Reliability+Doc).
- Data Products will see the Sentry error in #log-data-products-alerts and respond to it.

File: app/data_providers/slate_providers/slate_provider.py [hunk 4]
```diff
@@ -75,7 +76,12 @@ def configuration_id(self) -> str:
         """
         :return: UUID slate's configuration id, identifying the context and type of content that this slate provides.
         """
-        return str(uuid5(UUID(self.candidate_set_id), self.provider_name))
+        try:
+            # Try using the candidate_set_id as the uuid5 namespace.
+            return str(uuid5(UUID(self.candidate_set_id), self.provider_name))
+        except ValueError:
```
[the remark this answers is not in the document]
Comment (mmiermans) [reply app/data_providers/slate_providers/slate_provider.py:82]: From what I can tell, Python does not provide a way to detect whether a string is a UUID, other than to catch the `ValueError`.

An alternative idea would be to change the `candidate_set_id` type to `Union[UUID, str]` and have an if-condition on the type. That would risk a developer accidentally returning a Feature Group UUID as a string type, which would then result in a different `configuration_id` value.

A [try-except benchmark](https://stackoverflow.com/questions/2522005/cost-of-exception-handlers-in-python) suggests that catching an exception takes 0.2 ms versus 0.02ms for evaluating an if-condition with a single comparison.

What would you do?

[the remark this answers is not in the document]
Comment (mmiermans) [reply app/data_providers/slate_providers/slate_provider.py:82]: Maybe the right course of action would be to change the Snowplow/Snowflake schemas and move away from the `configuration_id` completely. We could emit `candidate_set_id` and `provider_class_name` from RecommendationAPI and report based on those.

That would be a bigger effort, so I wouldn't want to block this PR on it. I'd also prefer to wait with any significant schema changes until analytics has requirements defined for reporting, to avoid having to make schema changes twice.

[the remark this answers is not in the document]
Comment (mmiermans) [reply app/data_providers/slate_providers/slate_provider.py]: I added a function `candidate_set_uuid` which implementing classes (currently only NewTabSlateProvider) can override to return a UUID candidate_set_id. This way no try-except is necessary. I like this better than the try-except because it contains the logic to NewTabSlateProvider to convert a string to UUID.

File: app/data_providers/corpus/corpus_api_client.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+            corpus_items = []
+            for item in all_items:
+                corpus_item: CorpusItemModel = CorpusItemModel.parse_obj(item['corpusItem'])
+                corpus_items.append(corpus_item)
+                self._corpus_item_scheduled_date[corpus_item.id] = item['scheduledDate']
+
+        return corpus_items
+
+    def get_scheduled_date(self, corpus_item_id: str) -> str:
```
Comment (eu9ene) [inline app/data_providers/corpus/corpus_api_client.py:57]: Optional[str]?

**Review by eu9ene — APPROVED** [review]
